File: leetcode/17.LetterCombinationsOfAPhoneNumber.go
package leetcode

func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return []string{}
	}

	phone := map[string][]string{
		"2": {"a", "b", "c"},
		"3": {"d", "e", "f"},
		"4": {"g", "h", "i"},
		"5": {"j", "k", "l"},
		"6": {"m", "n", "o"},
		"7": {"p", "q", "r", "s"},
		"8": {"t", "u", "v"},
		"9": {"w", "x", "y", "z"},
	}

	var result []string
	var dfs func(i int, current string)

	dfs = func(i int, current string) {
		if len(current) == len(digits) {
			result = append(result, current)
		}

		if i >= len(digits) {
			return
		}

		digit := string(digits[i])
		letters, _ := phone[digit]

		for _, letter := range letters {
			dfs(i+1, current+letter)
		}
	}

	dfs(0, "")

	return result
}

File: leetcode/17.LetterCombinationsOfAPhoneNumber.ts
function letterCombinations(digits: string): string[] {
  const result: string[] = []
  const map: Record<number, string> = {
    2: 'abc',
    3: 'def',
    4: 'ghi',
    5: 'jkl',
    6: 'mno',
    7: 'pqrs',
    8: 'tuv',
    9: 'wxyz',
  }

  const dfs = (startIndex: number, path: string[]) => {
    if (startIndex === digits.length) {
      result.push(path.join(''))
      return
    }

    const words = map[Number(digits[startIndex])]
    for (const word of words) {
      path.push(word)
      dfs(startIndex + 1, path)
      path.pop()
    }
  }

  dfs(0, [])

  return result
}
console.log(letterCombinations('23'))
